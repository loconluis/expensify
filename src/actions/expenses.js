import uuid from 'uuid';
import database from '../firebase/firebase';

// Components calls action generator
// Action generator return object
// Component dispatches objects
// Redux store changes

// Components calls action generator
// action generator returns function
// component dispatches a function (?)
// function runs (has the ability  to dispatch other actions and do whatever it wants)

// ADD_EXPENSE,
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expensesData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expensesData;
    const expense = { description, note, amount, createdAt }

    return database.ref(`users/${uid}/expenses`).push(expense)
      .then((ref) => dispatch(addExpense({ id: ref.key, ...expense})))
  }
}

// REMOVE_EXPENSE,
export const removeExpense = ( { id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove()
      .then((ref) => dispatch(removeExpense({ id })));
  }
};

// EDIT_EXPENSE,
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates)
      .then(() => dispatch(editExpense(id, updates)));
  }
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).once('value')
      .then((snapshot) => {
        const expenses = []; // to clean when this promise run
        snapshot.forEach(childSnapshot => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        })
        return expenses // to get access to expenses on the next promise
      }).then((expenses) => dispatch(setExpenses(expenses)) )
  }
}