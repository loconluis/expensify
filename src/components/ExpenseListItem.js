import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';


export const ExpenseListItem = ({id, description, amount, createdAt}) => {
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>
        <strong>{numeral(amount / 100).format('$0,0.00')}</strong>
        - 
        {moment(createdAt).format('MMMM Do, YYYY')}
      </p>
    </div>
  )
};

export default ExpenseListItem