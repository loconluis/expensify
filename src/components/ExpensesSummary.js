import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';

import expensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';

export const ExpensesSummary = ({ expensesCount, total }) => {
  return (
    <div className="page-header">
      <div className="container">
        <h1 className="page-header__title">Viewing <span>{ expensesCount }</span> totalling <span>{ numeral(total / 100).format('$0,0.00')  }</span> </h1> 
        <div className="page-header__actions">
          <Link to="/create" className="btn">Add Expense</Link>
        </div>
      </div>
    </div>
  )
}

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expensesCount: visibleExpenses.length,
    total: expensesTotal(visibleExpenses)
  }
}



export default connect(mapStateToProps)(ExpensesSummary)