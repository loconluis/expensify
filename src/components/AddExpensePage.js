import React from 'react';
import { connect } from 'react-redux';

// component
import ExpenseForm from './ExpenseForm';
// action
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onSubmit(expense) {
    this.props.onSubmit(expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <h1>Add Expense:</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
    </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({ onSubmit: (expese) => dispatch(addExpense(expense)) })

export default connect(undefined, mapDispatchToProps)(AddExpensePage)