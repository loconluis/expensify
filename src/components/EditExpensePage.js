import React from 'react';
import { connect } from 'react-redux';

// Component
import ExpenseForm from './ExpenseForm';
// action
import { editExpense, removeExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);

    this.onSubmit = this.onSubmit.bind(this);
    this.onRemove = this.onRemove.bind(this);
  }

  onSubmit(expense) {
    // props.dispatch(editExpense(props.expense.id, expense))
    this.props.editExpense(this.props.expense.id, expense)
    this.props.history.push('/'); // Came back for the history
  }

  onRemove() {
    // props.dispatch(removeExpense({ id: props.expense.id }))
    this.props.startRemoveExpense({ id: this.props.expense.id })
    this.props.history.push('/'); // Came back for the history
  }

  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button
          onClick={this.onRemove}
        >Remove</button>
      </div>
    );
  };
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
