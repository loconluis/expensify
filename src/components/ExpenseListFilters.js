import React from 'react';
import { connect } from 'react-redux';

import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndtDate } from '../actions/filters'
// React-Dates
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      calendarFocused: null,
    }

    this.onDatesChange = this.onDatesChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
    this.onTextChange = this.onTextChange.bind(this);
    this.onSortChange = this.onSortChange.bind(this);
  }

  onDatesChange({ startDate, endDate }) {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }

  onFocusChange(calendarFocused) {
    this.setState(() => ({ calendarFocused }))
  }

  onTextChange (e) {
    this.props.setTextFilter(e.target.value)
  }

  onSortChange (e) {
    e.target.value === 'amount' ? this.props.sortByAmount() : this.props.sortByDate()
  }
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    )
  }
}

const mapStateToProps = (state) => ({ filters: state.filters });


const mapDispatchToProps = (dispatch, props) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndtDate: (endDate) => dispatch(setEndtDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);