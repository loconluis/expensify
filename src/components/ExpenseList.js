import React from 'react'
import { connect } from 'react-redux';

// Component
import ExpenseListItem from './ExpenseListItem';

// Selector
import selectExpenses from '../selectors/expenses';

const ExpenseList = (props) => {
  return (
    <div className="container">
      <div className="list-header">
        <div className="show-for-mobile">Expenses</div>
        <div className="show-for-desktop">Expense</div>
        <div className="show-for-desktop">Amount</div>
      </div>
      <div className="list-body">
        {
          props.expenses.length === 0 ? (<div className="list-item list-item--empty"><span>No Expenses</span></div>) : (props.expenses.map((expense, index) =><ExpenseListItem key={index} {...expense} />))
        }
      </div>
    </div>
  )
}

const mapStateToProps = state => (
  {
    expenses: selectExpenses(state.expenses, state.filters),
  }
);

export { ExpenseList as ExpenseListComponent };

export default connect(mapStateToProps)(ExpenseList);
