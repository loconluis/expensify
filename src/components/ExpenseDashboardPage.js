import React from 'react';
// componets
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters'

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;