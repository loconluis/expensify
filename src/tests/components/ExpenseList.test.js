import React from "react";
import { shallow } from 'enzyme';
// The component
import { ExpenseList } from '../../components/ExpenseList';
// Data seed
import expenses from '../fixtures/expenses';

test('should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />)
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList with empty message', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />)
  expect(wrapper).toMatchSnapshot();
});