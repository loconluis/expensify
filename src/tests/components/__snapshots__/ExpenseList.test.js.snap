// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div
  className="container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--empty"
    >
      <span>
        No Expenses
      </span>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div
  className="container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItem
      amount={100}
      createdAt={0}
      description="Gum"
      id="1"
      key="0"
      note=""
    />
    <ExpenseListItem
      amount={200}
      createdAt={-345600000}
      description="Choco Bar"
      id="2"
      key="1"
      note=""
    />
    <ExpenseListItem
      amount={456}
      createdAt={345600000}
      description="Clothing"
      id="3"
      key="2"
      note=""
    />
  </div>
</div>
`;
